Fixtures that allow install of a plugin from a target repo

DHaussermann opened the pull request:
This PR accomplishes two things
1. Implements reusable fixtures needed so you can easily download a plugin from a repo and enable it in ay Playwright test
2. Adds a test that downloads and enables a specific release of the demo plugin 

The next PR (if this is approved) will automate the demo plugin tests that I do manually each release

I realize this approach is maybe not ideal or is controversial as I did not add the test to the demo plugin repo. I am wiling to discard this later and  redo it in the demo plugin repo if needed. But for now this PR getting merged would get us into this position

**Cons:**
- Does not follow the convention of having Playwright test in the respective plugin repo
- Tests do not travel with the code as would be expected   
- The demo plugin test its self uses the 53 MB installer that has all the binaries. This is not great but I don't know how else to confirm it works locally AND on CI 
(If I recall, the plugin marketplace has a mechanism that determines your platform and only serves one binary maybe this can be leveraged in the future) 
- I'm sure there may be more :)

**Pros**:
- I can quickly get a working model that I can build on to write tests for the demo plugin
- The Demo plugin is unique in that it validates the behavior of the infrastructure used by all other plugins. In this respect, it is not illogical to tightly link these tests to the server repo rather than a self contained set of tests that can target any server build.
- Furthermore, when I do this test manually, I am looking for regressions in the webapp and the plugin API of the server. I would argue it is more the server that is under test than the plugin. As such this is a more reasonable implementation. 
- The follow-up PR will contain dozens of plugin webapp tests that will navigate the UI. All these page object expansions serve us best in the library here. To my understanding seeing how the Calls Playwright automation works, there would have been no shared benefit to the library by copying that implementation style   
- Being quickly able to download and enable a plugin with Playright will likely deliver future benefit

#### Release Note
```release-note
NONE
```

lieut-data commented on e2e-tests/playwright/lib/src/server/plugin.ts:
```diff
@@ -0,0 +1,29 @@
+    await client.enablePlugin(pluginId);
+
+    // Wait for activation
+    await new Promise((resolve) => setTimeout(resolve, 1000));
```
This is a bit arbitrary: should we try without and see how much tweaking is actually required in practice?

lieut-data commented on e2e-tests/playwright/specs/functional/plugins/demo_plugin_installation.spec.ts:
```diff
@@ -0,0 +1,33 @@
+        pw.mergeWithOnPremServerConfig({
+            FileSettings: {EnablePublicLink: true},
+        } as Partial<AdminConfig>),
+    );
```
Is there a `patchConfig` we can use instead? Ideally we don't have to care at the code-level about the "whole config" and just use the discrete endpoint instead.

lieut-data commented on e2e-tests/playwright/lib/src/server/plugin.ts:
```diff
@@ -0,0 +1,29 @@
+export async function verifyPluginActive(client: Client4, pluginId: string): Promise<boolean> {
+    const plugins = await client.getPlugins();
+    return plugins.active.some((plugin: PluginManifest) => plugin.id === pluginId);
+}
```
I'm guessing the `setTimeout` above were meant to allow this to succeed right away. Is there a PlayWright pattern we could use instead that would allow us to re-test this until it succeeds (subject to some timeout)?

saturninoabril commented on e2e-tests/playwright/lib/src/server/plugin.ts:
```diff
@@ -0,0 +1,29 @@
+import {Client4} from '@mattermost/client';
+import {PluginManifest} from '@mattermost/types/plugins';
+
+export async function installAndEnablePlugin(
```
Items to consider:
- If plugin is already active then no need to force install and activate given same version
- If plugin is already installed, then just enable
- Should cater installation by URL or file

Still more to improve but would be good to port from https://github.com/mattermost/mattermost/blob/master/e2e-tests/cypress/tests/support/api/plugin.js#L57

saturninoabril commented on e2e-tests/playwright/specs/functional/plugins/demo_plugin_installation.spec.ts:
```diff
@@ -0,0 +1,33 @@
+    // Optional: Get plugin details
+    const plugins = await adminClient.getPlugins();
+    const demoPlugin = plugins.active.find((p) => p.id === 'com.mattermost.demo-plugin');
+    expect(demoPlugin).toBeDefined();
```
This is more of an API/client test. The first part with client usage makes sense for data state/preparation but on the last part, I'd suggest to add UI action and verification for validating that the demo plugin is installed and active, could be by doing slash command and its output.

DHaussermann commented:
Changes Made:

1. Removed 2 waits from plugin.ts that were not needed for the test to function
   - File: e2e-tests/playwright/lib/src/server/plugin.ts
   - Removed: 5-second wait after installPluginFromUrl (line 16-17)
   - Removed: 1-second wait after enablePlugin (line 22-23)
   - Reason: Arbitrary timeouts are unreliable; tests should validate actual state

2. Replaced config merge operation with more conventional patch edit to set config value
   - File: e2e-tests/playwright/specs/functional/plugins/demo_plugin_installation.spec.ts
   - Replaced: adminClient.updateConfig(pw.mergeWithOnPremServerConfig({...}))
   - With: adminClient.patchConfig({FileSettings: {EnablePublicLink: true}})
   - Reason: patchConfig is cleaner, only updates specific fields, no merge needed
3. Added retry pattern with expect.poll() for plugin activation check  
   - Replaced: Direct await with expect(isActive).toBe(true)
   - With: await expect.poll(async () => { return await pw.isPluginActive(...) }).toBe(true)                                                                                                                                                                                                                                                                  
4. Renamed `verifyPluginActive` → `isPluginActive` as the function does not perform a vberification in the technical sense
5. Expanded capabilities of `installAndEnablePlugin` by creating and using 2 other function for `isPluginActive` and `getPluginStatus` 
6. I added a test to verify the plugin in the UI by running a slash command but during `await channelsPage.postMessage('/demo_plugin true');` as the input box is in use the overlay with onboarding list appears again and I cannot continue to use the UI.
@saturninoabril , do we have a known solution to this? My work is not done as my test will not pass due the overlay. Surely there is a standard way o solve this problem. 
`EnableOnboardingFlow: false,` did not work (possibly a restart is needed after patchig)
Dismiss with escape key did not work as the modal appears when the input box get's key stokes


**note**: I did not implement support for uploading a plugin from file instead of URL. I started to work on this after review of the Cypress model but it became complex. I could not just retrofit my function to accept a file path. The `client.uploadPlugin()` available as part of `Client4` expects a file object. This is challenging as I'm not working in a browser. File upload support would require additional conversion logic and is better suited for a separate PR focused on UI-based upload tests.

saturninoabril commented:
Thanks for the updates Dylan.

Another way is to disable the user preference. To apply globally, you update the initSetup (init.ts) with the last two entries:
```ts
// Update user preference
const preferences: PreferenceType[] = [
    {user_id: user.id, category: 'tutorial_step', name: user.id, value: '999'},
    {user_id: user.id, category: 'crt_thread_pane_step', name: user.id, value: '999'},
    {user_id: user.id, category: 'onboarding_task_list', name: 'onboarding_task_list_show', value: 'false'},
    {user_id: user.id, category: 'onboarding_task_list', name: 'onboarding_task_list_close', value: 'false'},
];
```

For file upload, I'm good as is to consider upload by URL for now.

saturninoabril commented:
PR submitted: https://github.com/mattermost/mattermost/pull/34912
